### ISSUE TITLE
`calc` builtin conflicts with commonly-used tool

### ISSUE THREAD

**Issue by jbg** [redox-os/ion#542]
**Reproduction**: Run `calc`

**Expected behavior**: The common "C-style arbitrary precision calculator", distributed as a part of *NIX distributions since 1994, launches.

**Actual behavior**: Apparently my shell is hung, but in fact the shell builtin is waiting for input and does not respond to Ctrl-C.

Maybe I'm one of the last people still using the fantastic `calc` calculator, but I hope not. Please consider renaming your builtin, or if not, at least make it respond to Ctrl-C :)

**Comment by mmstick** [comment]
I've actually never seen that distributed with any Linux distributions. Anyway, if you want to ensure that you run that one, you can simply call it directly with `/usr/bin/calc`, but I will add a means of responding to Ctrl+C.

**Comment by jbg** [comment]
Yes, I know I can call `/usr/bin/calc`. `calc` is distributed with at least Ubuntu, Debian, Arch and Red Hat, as well as in Homebrew and MacPorts on macOS. I think sometimes the Debian-based distros call the package `apcalc`, but the installed binary is still called `calc`. Anyway, support for the interrupt signal will be a welcome improvement. Is there any alias mechanism or any other way I can override the built-in without having to type the full path to `calc`?

**Comment by jackpot51** [comment]
I think it would be good if the `calc` builtin was renamed. `Bash` has `$((1 + 1))`, for example. I think this would be a good thing for `ion` as well. The calc command prompt in ion could also be removed, as it is broken compared to the one in extrautils.

**Comment by jbg** [comment]
fwiw, it seems `ion` already supports the `bash`-like syntax for calculations:

```
jbg:~# echo $((1 + 1))
2
```

**Comment by jackpot51** [comment]
Cool

**Comment by mmstick** [comment]
@jbg Can't find it packaged on Solus, Arch and Fedora seems to offer it in a `calc` package though. Not installed by default. There is an `alias` builtin that you can use, but just know that calling an external program to perform calculations will be much more costly than using the built-in calc.

**Comment by mmstick** [comment]
Also, quitting the built-in calc is done by pressing Ctrl + D.

**Comment by jbg** [comment]
Thanks @mmstick, I will use `alias`. I did already figure out that it quits once it gets EOF on input. I know compliance with existing standards isn't a stated goal of the project, but given that `ion` and therefore the built-in `calc` can run on Linux, it would be nice if it was a well-behaved Linux program when it did so, which would include responding to SIGINT.

`calc` is not installed by default on any distro I'm aware of, but I've never seen it not packaged on a distro (hadn't heard of Solus before today though). Many other tools are not installed by default on many distributions, but would be problematic for certain groups of users if their names were shadowed by the shell. For example, `gcc`.

I guess the question of why I want to use the standalone `calc` program instead of the built-in is a bit out of the context of this issue, but since you brought up the extra cost of running the program...

I don't mind running an executable to run a calculator (this is not for scripting, it's for interactive use). The overhead of launching an executable is not a problem when I will be using it for several minutes or longer and is also justified by the fact that `calc` is significantly more capable than the built-in calc.

I couldn't find much documentation about the built-in `calc` program, but from some basic tests (see below) it is using floating-point arithmetic, and silently truncates values. This is not suitable for my purposes (and in my opinion less than ideal for use as a general-purpose interactive calculator, though perfectly fine for evaluating expressions in a programming/shell language).

```
jbg:~# calc
10.00000000000000000001
[]> 10
0.1*0.1
[]> 0.010000000000000002
```

Compare to:

```
jbg:~# /usr/bin/calc
C-style arbitrary precision calculator (version 2.12.6.0)
Calc is open software. For license details type:  help copyright
[Type "exit" to exit, or "help" for help.]
; 10.00000000000000000001
        10.00000000000000000001
; 0.1*0.1
        0.01
```

**Comment by mmstick** [comment]
@jbg So information on `calc` can be found from the [calculate](https://github.com/redox-os/calc) crate, which is where all the logic is. The code base is pretty small, and relatively easy to read. I suspect the reason for the differences in values between the built-in and the external calc commands are likely that the external calc is using a different algorithm for performing it's calculations, either intentionally in the source code, or by differences between LLVM and GCC and their respective compiler optimization levels.

- The builtin calc supports being supplied mathematical expressions in both prefix (Polish) notation, and the standard infix notation. `calc '+ * + 4 2 3 * 3 2'` will return `24`, for example.
- Values are parsed and stored as either a signed 64-bit integer (noted to be either decimal or hexadecimal), or 64-bit floating point value, depending on what the expression demands.
- Therefore, floats aren't used if floats aren't required -- integer math is the default choice.
- It's not implemented, but 128-bit integers could be used if there's a demand for it, as Rust natively supports `i128` and `f128` types. If you would like to see 128-bit floating point precision, I could easily look into integrating it.
- All mathematical operations are using standard operators in Rust, so there's no truncation being performed.
- [If truncation were to happen, it would be treated as an error](https://github.com/redox-os/calc/blob/master/src/error.rs#L69).
- You'll notice though that you can supply an input like `calc 5.23234 / 13.2323` and get back `0.3954218087558474`, which is rather precise.

Supported mathematical operations at the moment are:

- Plus (+)
- Minus (-)
- Divide (/)
- Multiply (*)
- Exponent (**)
- Square (N²)
- Cube (N³)
- BitWiseAnd (&)
- BitWiseOr (|)
- BitWiseXor (^)
- BitWiseNot (~N)
- BitWiseRShift (>>)
- BitWiseLShift (<<)
- Modulo (%)
- Parenthesis (not required with Polish notation)

**Comment by mmstick** [comment]
And, the general reason for not quitting the command when pressing Ctrl+C is because REPLs on Linux normally use Ctrl+C to clear inputs. Ctrl + D is for qutting REPLs.

**Comment by jbg** [comment]
Thanks for the info.

The reason for the difference between the results is actually that `calc` (the program, not the built-in) is not a floating-point calculator (this is made clear in its package name on some distros - `apcalc` for arbitrary precision calculator). In the sciences a lot of people use this program because it can remain accurate to any precision and does not exhibit the error that floating point operations do.

I understand not using Ctrl-C to quit a REPL. I guess it's just a bit problematic that I expected to run a reasonably well-known (at least in my field) program, got a different program, and didn't realise because the built-in doesn't produce any output when it's run interactively. That combined with the lack of response to SIGINT makes it "feel" like the shell has hung rather than running the command I expected it to. Slightly more problematic is that if I had not noticed the lack of output and started making calculations, I would have inaccuracies in my results because instead of running the arbitrary precision calculator I intended to run, I got a floating-point calculator just because I was using a different shell.

Anyway, I think I've made clear that I think the naming is unfortunate, but I can certainly understand the desire to name a built-in calculator `calc`. If it printed some header/prompt when run interactively I think it would be more obvious that you have run a program, in which case not supporting SIGINT seems okay (but maybe it's worth printing something, e.g. like the Python REPL does, when receiving SIGINT).

**Comment by mmstick** [comment]
The [decimal](https://github.com/alkis/decimal) crate may be of interest, as it offers 128-bit decimal float support (a float backed by integer arithmetic).

**Comment by mmstick** [comment]
So I have a PR for `calculate` to use `d128` instead of `f64`, to eliminate usage of FPUs and rounding errors that are inherent with their usage, and to simultaneously increase precision, up to 34 decimal places: https://github.com/redox-os/calc/pull/18

It's not arbitrary precision, but it's a step in that direction.

**Comment by mmstick** [comment]
Alright, so I've switched Ion to using my fork of `calculate` with the `decimal` branch that contains the usage of `d128` as a replacement for `f64`. Preforming any kind of math with decimals should now be precise up to 34 decimals.

```
mmstick:~/S/ion# calc
ion-calc: 10.00000000000000000001
10.00000000000000000001
ion-calc: 0.1 * 0.1
0.01
ion-calc: exit
mmstick:~/S/ion# 
```

**Comment by jbg** [comment]
very, very nice!

**Comment by BafDyce** [comment]
(Disclaimer: I haven't read the entire issue, so sorry if this is already mentioned somewhere)

We could also add the possibility to let the user disable/rename certain builtins.

**Comment by bb010g** [comment]
I think renaming it to `icalc` isn't that bad of an idea. Zsh's builtin calculator interface is [`zcalc`](http://jlk.fjfi.cvut.cz/arch/manpages/man/extra/zsh/zshcontrib.1.en).

**Comment by mmstick** [comment]
Would everyone else be okay with `calc` being renamed to `icalc`?
